sae_lens/sae.py

```diff
@@ -390,109 +392,23 @@ def forward(
                 # This is in a no_grad context to detach the error, so we can compute SAE feature gradients (eg for attribution patching). See A.3 in https://arxiv.org/pdf/2403.19647.pdf for more detail
                 # NOTE: we can't just use `sae_error = input - x_reconstruct.detach()` or something simpler, since this would mean intervening on features would mean ablating features still results in perfect reconstruction.
-
-                # move x to correct dtype
-                x = x.to(self.dtype)
-
-                # handle hook z reshaping if needed.
-                sae_in = self.reshape_fn_in(x)  # type: ignore
-
-                # handle run time activation normalization if needed
-                sae_in = self.run_time_activation_norm_fn_in(sae_in)
-
-                # apply b_dec_to_input if using that method.
-                sae_in_cent = sae_in - (self.b_dec * self.cfg.apply_b_dec_to_input)
-
-                # "... d_in, d_in d_sae -> ... d_sae",
-                hidden_pre = sae_in_cent @ self.W_enc + self.b_enc
-                feature_acts = self.activation_fn(hidden_pre)
-                x_reconstruct_clean = self.reshape_fn_out(
-                    self.apply_finetuning_scaling_factor(feature_acts) @ self.W_dec
-                    + self.b_dec,
-                    d_head=self.d_head,
-                )
-
-                sae_out = self.run_time_activation_norm_fn_out(sae_out)
-                sae_error = self.hook_sae_error(x - x_reconstruct_clean)
-            return self.hook_sae_output(sae_out + sae_error)
-
-        # TODO: Add tests
-        elif self.use_error_term and self.cfg.architecture == "gated":
-            with torch.no_grad():
-                x = x.to(self.dtype)
-                sae_in = self.reshape_fn_in(x)  # type: ignore
-
-                # handle run time activation normalization if needed
-                sae_in = self.run_time_activation_norm_fn_in(sae_in)
-
-                # apply b_dec_to_input if using that method.
-                sae_in = sae_in - (self.b_dec * self.cfg.apply_b_dec_to_input)
-
-                gating_pre_activation = sae_in @ self.W_enc + self.b_gate
-                active_features = (gating_pre_activation > 0).float()
-
-                # Magnitude path with weight sharing
-                magnitude_pre_activation = self.hook_sae_acts_pre(
-                    sae_in @ (self.W_enc * self.r_mag.exp()) + self.b_mag
-                )
-                feature_magnitudes = self.activation_fn(magnitude_pre_activation)
-                feature_acts_clean = self.hook_sae_acts_post(
-                    active_features * feature_magnitudes
-                )
-                x_reconstruct_clean = self.reshape_fn_out(
-                    self.apply_finetuning_scaling_factor(feature_acts_clean)
-                    @ self.W_dec
-                    + self.b_dec,
-                    d_head=self.d_head,
-                )
```

Comment:
I assumed the code in these codepaths is trying to exactly recreate all the `encode_x()` variants followed by `decode()`, but is duplicating code just to avoid triggering hooks.

Reply:
yes. Disabling hooks is a good solution, better than duplicating code.